CUDA Vectorized Dropout

slayton58 opened the pull request:
Add vectorization to dropout kernels for both reads & writes. Moved the `masked_scale_kernel` implementation to `TensorIterator` to pick up recent autovectorization additions by @zasdfgbnm , and wrote a vectorized specialization of the dropout training kernel (along with some fairly conservative dispatch logic).

slayton58 committed:
Vectorize dropout (training & inference)

Maintains legacy path by @ngimel for non-vectorisable case(s), uses
fairly conservative is-vectorizable logic:
* all inputs & outputs contiguous
* No remainder (t.numel() % vector_size == 0)

slayton58 committed:
Fix lambda signature & iter options

Now passes test_nn.py

slayton58 committed:
Incorporate suggestions from @zasdfgbnm

ngimel commented on aten/src/ATen/native/cuda/Dropout.cu:
```diff
@@ -16,8 +20,84 @@ namespace {
+    *value = *reinterpret_cast<LoadT*>(&a.data[aOffset]);
+
+    scalar_t r[VEC];
+    uint8_t mask[VEC];
```
bool tensors are now supported, so you could use it to store mask, but up to you.

ngimel commented on aten/src/ATen/native/cuda/Dropout.cu:
```diff
@@ -16,8 +20,84 @@ namespace {
+
+    //curand_uniform_double was pure evil anyway, not doing what it promises, and there's nothing for halfs, so generate float for everything
+    // Note: need a new set of random values per 4 elements -- we'll handle VEC elements in this thread, so need ceil(VEC / 4)
+    // sets of rand.
```
nit: static_assert VEC<=4 will make it safer to future improvements

slayton58 replied:
👍

ngimel commented on aten/src/ATen/native/cuda/Dropout.cu:
```diff
@@ -117,12 +227,29 @@ fused_dropout_cuda(const Tensor& self, double p, Generator * gen_){
+
+      if (vec_size > 1) {
+        switch (vec_size) {
+         case 8:
```
What's that? You don't generate enough randoms to handle that (or I'm blind)

slayton58 replied:
From side-conversation, this was vestigial code that I forgot to remove - `vec_size=8` is not supported for now.

ngimel commented:
Windows build failure is real.

slayton58 committed:
Address review by @ngimel

slayton58 committed:
Address win compile error

slayton58 committed:
Fix for #34085

Use the layout suggested by the input tensor for the output instead of
legacy

slayton58 committed:
Vectorize in channels_last case

Can now vectorize as long as input (self tensor) is contiguous based on
the format returned by `self.suggested_memory_format()`. Previously
.contiguous() was used, which would return `False` on non-NCHW tensors.
